### PR TITLE
Wip rgw cksum

### PR DESCRIPTION
**Pull request by mattbenjamin**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

File: src/rgw/rgw_crc_digest.h [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    Crc32c() { Restart(); }
+
+    void Restart() { crc = initial_value; }
+
+    void Update(const unsigned char *data, uint64_t len) {
+      crc = ceph_crc32c(crc, data, len);
+    }
+
+    void Final(unsigned char* digest) {
```
Comment (mkogan1) [inline src/rgw/rgw_crc_digest.h:85]: cont. from slack conversation, to follow consistent byte ordering as in CRC32 suggesting to add `byteswap()` in CRC32C also:
```
    void Final(unsigned char* digest) {
      uint32_t final = crc ^ 0xffffffff;
      if constexpr (std::endian::native != std::endian::big) {
	final = rgw::digest::byteswap(final);
      }
      memcpy((char*) digest, &final, sizeof(final));
    }
  }; /* Crc32c */
  ```

Comment (mattbenjamin) [reply src/rgw/rgw_crc_digest.h:85]: well iiuc, what matters is that it matches a checksum generated by an aws toolkit during upload;  I haven't actually checked that yet, but what we're doing now matches the table of outputs on crccalc.org, correct?  I will check using AWS toolkit for golang later today.

File: src/test/rgw/test_rgw_cksum.cc [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+  DigestVariant dv = rgw::cksum::digest_factory(t);
+  Digest* digest = get_digest(dv);
+
+  ASSERT_NE(digest, nullptr);
+
+  digest->Update((const unsigned char *)dolor.c_str(), dolor.length());
+
+  auto cksum = rgw::cksum::finalize_digest(digest, t);
+  /* compare w/known value https://crccalc.com/ */
+  ASSERT_EQ(cksum.hex(), "4b2edc95");
```
Comment (mkogan1) [inline src/test/rgw/test_rgw_cksum.cc:96]: following previous comment regarding `byteswap()` [1]
suggesting a change to reverse the byteorder: 
`ASSERT_EQ(cksum.hex(), "95DC2E4B");`
to be consistent with the display in:
https://crccalc.com/?crc=Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.&method=crc32&datatype=ascii&outtype=0
image


[1] https://github.com/ceph/ceph/pull/54584/files#r1402416023

Comment (mattbenjamin) [reply src/test/rgw/test_rgw_cksum.cc:96]: oh, I guess I reversed it myself!

Comment (mattbenjamin) [reply src/test/rgw/test_rgw_cksum.cc:96]: I thought the current code was doing that already without byteswap

Comment (mattbenjamin) [reply src/test/rgw/test_rgw_cksum.cc:96]: will recheck and fix

Comment (mattbenjamin) [reply src/test/rgw/test_rgw_cksum.cc:96]: my intent was to match that value, which you also produced with crc32 and cksfv

**Comment by mattbenjamin** [comment]
and that's also not correct, that was using crc32

**Comment by mattbenjamin** [comment]
to start with, matching the published value at crccaclc.org seems appropriate

File: src/test/rgw/test_rgw_cksum.cc [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+  Digest* digest = get_digest(dv);
+
+  ASSERT_NE(digest, nullptr);
+
+  digest->Update((const unsigned char *)dolor.c_str(), dolor.length());
+
+  auto cksum = rgw::cksum::finalize_digest(digest, t);
+  
+  /* compare w/known value https://crccalc.com/ */
+  ASSERT_EQ(cksum.hex(), "98b2c5bd" /* not quite bdc5b298 */);
```
Comment (mkogan1) [inline src/test/rgw/test_rgw_cksum.cc:79]: ack, the CRC32 (without C) value is already consistent (has `byteswap()`) per https://crccalc.com/
image

File: src/test/rgw/CMakeLists.txt [hunk 4]
```diff
@@ -246,11 +246,13 @@ target_link_libraries(unittest_rgw_lc
 
 # unittest_rgw_cksum
 add_executable(unittest_rgw_cksum test_rgw_cksum.cc)
-add_ceph_unittest(unittest_rgw_cksum)
```
[the remark this answers is not in the document]
Comment (mattbenjamin) [reply src/test/rgw/CMakeLists.txt]: sorry, basically yes (restored)

**Pull request closed without merging**